src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,205 @@
+#include "scheduling_alg.h"
+using namespace std;
+
+ClusterResourceScheduler::ClusterResourceScheduler(int64_t local_node_id, const NodeResources &local_node_resources) {
+  local_node_id_ = local_node_id;
+  AddOrUpdateNode(local_node_id, local_node_resources);
+}
+
+void ClusterResourceScheduler::UpdatePredefinedResources(NodeResources& old_resources,
+  const NodeResources& new_resources) {
+
+    // First, update its predefined resources.
+    for (int i = 0; i < PredefinedResources_MAX; i++) {
+      old_resources.capacities[i].total = new_resources.capacities[i].total;
+      old_resources.capacities[i].available = new_resources.capacities[i].available;
+    }
+}
+
+void ClusterResourceScheduler::UpdateCustomResources(
+  unordered_map<int64_t, ResourceCapacity>& old_custom_resources,
+  const unordered_map<int64_t, ResourceCapacity>& new_custom_resources) {
+
+  // Then, delete existing custom resources and replace them with the new ones.
```

Comment:
```suggestion
  // Delete existing custom resources and replace them with the new ones.
```
This comment probably isn't necessary anymore given that it's a small function.